File: 1-Part/5-chapter/quicksort.py
def quicksort(items, left=None, right=None):
    if left is None:
        left = 0
    if right is None:
        right = len(items) - 1
    print('\nquicksort() called on this range:', items[left:right + 1])
    print('................ The full list is:', items)

    if right <= left:
        return

    i = left
    pivotValue = items[right]
    print('.............. The pivot is:', pivotValue)

    for j in range(left, right):
        if items[j] < pivotValue:
            items[i], items[j] = items[j], items[i]
            i += 1
    items[i], items[right] = items[right], items[i]

    print('... After swapping, the range is:', items[left:right + 1])
    print('Recursively calling quicksort on:', items[left:i], 'and', items[i + 1:right + 1])

    quicksort(items, left, i - 1)
    quicksort(items, i + 1, right)


myList = [0, 5, 9, 4, 8, 2, 7, 6, 1]
quicksort(myList)
print(myList)

File: 1-Part/5-chapter/quicksort.js
function quicksort(items, left, right) {
    if (left === undefined) {
        left = 0
    }
    if (right === undefined) {
        right = items.length - 1
    }

    console.log(`quicksort() called on this range: ${items.slice(left, right + 1).join(', ')}`)
    console.log(`........... The full list is: ${items.join(', ')}`)

    if (right <= left) {
        return
    }

    let i = left
    let pivotValue = items[right]
    console.log(`....... The pivot is: ${pivotValue.toString()}`)

    for (let j = left; j < right; j++) {
        if (items[j] <= pivotValue) {
            [items[i], items[j]] = [items[j], items[i]]
            i++
        }
    }
    [items[i], items[right]] = [items[right], items[i]]
    console.log(`........ After swapping, the range is: ${items.slice(left, right + 1).join(', ')}`)
    console.log(`Recursively calling quicksort on: ${items.slice(left, i).join(', ')} and ${items.slice(i + 1, right + 1).join(', ')}`)

    quicksort(items, left, i -1)
    quicksort(items, i+1, right)
}

let myList = [0, 5, 9, 4, 8, 2, 7, 6, 1]
quicksort(myList)
console.log(myList.join(', '))